### PR TITLE
Make markdown spell-checking a Pester test

### LINKED ISSUE #8078: The Markdown spelling tests should be converted to a Pester test
Current tests:
https://github.com/PowerShell/PowerShell/blob/1dc4a7903f6a579640d3cf9e7104512e85b4455c/.vsts-ci/spelling.yml#L27

Example of the markdown lint tests written in pester:  
https://github.com/PowerShell/PowerShell/tree/master/test/common/markdown

### PR DESCRIPTION
**Pull request by RDIL**
# PR Summary
fixes #8078 
todo:
update contributing md

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [X] Issue filed:  #8078 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: test/common/markdown/markdown-link.tests.ps1 [hunk 1]
```diff
@@ -72,7 +70,7 @@ Describe "Verify Markdown Links" {
                 foreach ($failure in $failures) {
                     if($failure -like 'https://www.amazon.com*')
                     {
-                        # In testing amazon links often failed when they are valid
+                        # In testing Amazon links often failed when they are valid
```
Comment (RDIL) [inline test/common/markdown/markdown-link.tests.ps1]: @TravisEz13 why do Amazon links do this?

**Comment by RDIL** [comment]
@iSazonov @vexx32 can you two please take a look?  Need a second eye as I can't find my mistake.

**Comment by iSazonov** [comment]
First we need confirmation from @TravisEz13 - I am not sure that it is right direction for spelling.

**Comment by RDIL** [comment]
@iSazonov see issue #8078

File: .vsts-ci/linux.yml [hunk 2]
```diff
@@ -48,7 +48,7 @@ jobs:
 
   - powershell: |
       Get-ChildItem -Path env:
-    displayName: Capture environment
+    displayName: Capture Environment
```
Comment (iSazonov) [inline .vsts-ci/linux.yml]: @RDIL Before we continue please move all style changes to another PR. Thanks.

Comment (RDIL) [reply .vsts-ci/linux.yml]: done

Comment (iSazonov) [reply .vsts-ci/linux.yml]: I still see many style changes. Please revert all.

Comment (RDIL) [reply .vsts-ci/linux.yml]: That is all the style issues, rest are functionality changes

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        $job = $jobs.$key
+        $results = Receive-Job -Job $job -Wait
+        Remove-job -job $Job
+        foreach($jobResult in $results)
+        {
+            $file = $jobResult.file
+            $result = $jobResult.results
+            Context "Verify spelling in $file" {
+                $failures = $result -like '*spelling errors found in*'
```
Comment (vexx32) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
                $failures = @($result) -like '*spelling errors found in*'
```
If you don't force this as an array and there is only one or zero items in $result, it will return $false, which is one of the symptoms of the error (loop is passing `$false` as a value through to the pester test which is throwing).

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                $passes = $result -like '*free of spelling*'
```
Comment (vexx32) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
                $passes = @($result) -like '*free of spelling*'
```
Same thing here.

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    $groups = Get-ChildItem -Path "$PSScriptRoot\..\..\..\*.md" -Recurse | Where-Object {$_.DirectoryName -notlike '*node_modules*'} | Group-Object -Property directory
+
+    $jobs = @{}
+    # Start all spelling verification in parallel
+    Foreach($group in $groups)
+    {
+        Write-Verbose -verbose "starting jobs for $($group.Name) ..."
+        $job = Start-ThreadJob {
+            param([object] $group)
+            foreach($file in $group.Group)
```
Comment (vexx32) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
            foreach($file in $group)
```
I would just pass the group in as the array rather than the whole group object. See above and below.

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    $groups = Get-ChildItem -Path "$PSScriptRoot\..\..\..\*.md" -Recurse | Where-Object {$_.DirectoryName -notlike '*node_modules*'} | Group-Object -Property directory
+
+    $jobs = @{}
+    # Start all spelling verification in parallel
+    Foreach($group in $groups)
+    {
+        Write-Verbose -verbose "starting jobs for $($group.Name) ..."
+        $job = Start-ThreadJob {
+            param([object] $group)
```
Comment (vexx32) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
            param([object[]] $group)
```

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            foreach($file in $group.Group)
+            {
+                $results = mdspell --en-us --ignore-numbers --ignore-acronyms --report $file 2>&1
+                Write-Output ([PSCustomObject]@{
+                    file = $file
+                    results = $results
+                })
+            }
+        } -ArgumentList @($group)
```
Comment (vexx32) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
        } -ArgumentList @($group.Group)
```

File: .vsts-ci/misc-analysis.yml [hunk 8]
```diff
@@ -33,24 +33,22 @@ phases:
     displayName: Capture environment
     condition: succeededOrFailed()
 
-  - powershell:  Install-module pester -Scope CurrentUser -Force -SkipPublisherCheck
+  - powershell:  Install-Module -Name pester -Scope CurrentUser -Force
```
Comment (iSazonov) [inline .vsts-ci/misc-analysis.yml]: Revert style.

File: .vsts-ci/misc-analysis.yml [hunk 9]
```diff
@@ -33,24 +33,22 @@ phases:
...
   - powershell:  Write-Host "##vso[build.updatebuildnumber]$env:BUILD_SOURCEBRANCHNAME-$env:BUILD_SOURCEVERSION-$((get-date).ToString("yyyyMMddhhmmss"))"
     displayName: Set Build Name for Non-PR
     condition: ne(variables['Build.Reason'], 'PullRequest')
 
+  # Sometimes this folder is left behind with root permissions and causes many problems
   - bash: |
-      mdspell '**/*.md' '!**/Pester/**/*.md' --ignore-numbers --ignore-acronyms --report --en-us;
-    displayName: Test Spelling
+      sudo rm -rf ~/.npm/_cacache
+    displayName: Remove Old NPM Caches
```
Comment (iSazonov) [inline .vsts-ci/misc-analysis.yml]: Revert style.

File: .vsts-ci/misc-analysis.yml [hunk 10]
```diff
@@ -60,7 +58,7 @@ phases:
       Write-Host "##vso[results.publish type=NUnit;mergeResults=true;runTitle=Common Tests;publishRunAttachments=true;resultFiles=$path;]"
       if($results.TotalCount -eq 0 -or $results.FailedCount -gt 0)
       {
-        throw "markdown link tests failed"
+        throw "Markdown tests failed"
```
Comment (iSazonov) [inline .vsts-ci/misc-analysis.yml]: Revert style.

File: test/common/markdown/markdown-link.tests.ps1 [hunk 11]
```diff
@@ -121,7 +116,7 @@ Describe "Verify Markdown Links" {
 
                 if($verifyFailures)
                 {
-                    it "<url> should work" -TestCases $verifyFailures -Pending  {
+                    it "<url> should work" -TestCases $verifyFailures -Pending {
```
Comment (iSazonov) [inline test/common/markdown/markdown-link.tests.ps1]: Revert style.

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        $jobs.add($group.name,$job)
+    }
+
+    # Get the results and verify
+    foreach($key in $jobs.keys)
```
Comment (vexx32) [inline test/common/markdown/markdown-spell.tests.ps1]: This is precarious. If there is ever a folder named `keys` in the repo, that will break this logic. A hashtable with a key called `keys` will return the value of that key instead of all the keys in the hashtable. I do wish they'd fix that, but... 🤷‍♂️ 

Instead:
```suggestion
    foreach($key in $jobs.GetEnumerator().Name)
```

File: .vsts-ci/misc-analysis.yml (hunk 9 above)
Comment (RDIL) [reply .vsts-ci/misc-analysis.yml]: This isn't style, had to do it to make it a pester test...

File: .vsts-ci/misc-analysis.yml (hunk 10 above)
Comment (RDIL) [reply .vsts-ci/misc-analysis.yml]: same here, now this is both links and spelling (not style)

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        Remove-job -job $Job
+        foreach($jobResult in $results)
+        {
+            $file = $jobResult.file
+            $result = $jobResult.results
+            Context "Verify spelling in $file" {
+                $failures = $result -like '*spelling errors found in*'
+                $passes = $result -like '*free of spelling*'
+ 
+                $trueFailures = @()
```
Comment (vexx32) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
                $trueFailures = foreach ($Failure in $Failures) {
                    @{ Spell = $Failure }
                }
```

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+                foreach ($failure in $failures) {
```
Comment (vexx32) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion

```

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+                    $trueFailures += @{spell = $failure}
```
Comment (vexx32) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion

```

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+                }
```
Comment (vexx32) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
```

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+
+                # must have some code in the test for it to pass
+                function noop {}
+
+                if($passes) {
```
Comment (vexx32) [inline test/common/markdown/markdown-spell.tests.ps1]: Do we really need to execute an `It` block for any of these? Since we're doing `noop` anyway, this basically would just take execution time to do nothing. May as well simply not bother outputting passing values, almost. 🤔

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+        # Cleanup jobs for reliability
+        get-job | remove-job -force
+    }
+
+    AfterAll {
+        # Cleanup jobs to leave the process the same
+        get-job | remove-job -force
+    }
+
+    $groups = Get-ChildItem -Path "$PSScriptRoot\..\..\..\*.md" -Recurse | Where-Object {$_.DirectoryName -notlike '*node_modules*'} | Group-Object -Property directory
```
Comment (vexx32) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
    $groups = Get-ChildItem -Path "$PSScriptRoot\..\..\..\*.md" -Recurse | 
        Where-Object DirectoryName -notlike '*node_modules*' | 
        Group-Object -Property directory
```

File: test/common/markdown/markdown-link.tests.ps1 (hunk 11 above)
Comment (RDIL) [reply test/common/markdown/markdown-link.tests.ps1]: fixed

**Review by vexx32 — COMMENTED** [review]
Not entirely sure where the issue is _completely_ but this should at least help us figure out what's up with the output there.

File: test/common/markdown/markdown-spell.tests.ps1 (hunk 17 above)
Comment (RDIL) [reply test/common/markdown/markdown-spell.tests.ps1]: Good point.  Thanks for reviewing!

**Comment by RDIL** [comment]
I love pester 🤣 
```
Context Verify spelling in /home/vsts/work/1/s/tools/performance/README.md
      [-] >> 8 spelling errors found in 1 file should work 2ms
        RuntimeException: Tool reported spelling as wrong.
```

**Comment by vexx32** [comment]
Does `mdspell` allow us to see each spelling error @RDIL?

Might need to rethink how you're categorising and parsing the output 😉

**Comment by RDIL** [comment]
Yeah it does but I sadly don't have that level of expertise in powershell :(

**Comment by RDIL** [comment]
Thank you for all your help @vexx32!

**Comment by vexx32** [comment]
@rdil can you show me some example output from mdspell? I can probably figure out a regex to get the misspelled words.

**Comment by RDIL** [comment]
@vexx32 here:
https://cirrus-ci.com/task/5949319808024576
make sure to click on main to open the logs
also its at the very bottom of logs

**Comment by vexx32** [comment]
Ah, I see. So currently you're checking your final output line, the summary. OK... OK... hm. Can you mock up an example where there's spelling errors in at least two files for me? I can definitely regex that, just don't wanna miss anything. ❤️

**Comment by RDIL** [comment]
sure one min

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+    foreach($key in $jobs.GetEnumerator().Name)
+    {
+        $job = $jobs.$key
+        $results = Receive-Job -Job $job -Wait
+        Remove-job -job $Job
+        foreach($jobResult in $results)
+        {
+            $file = $jobResult.file
+            $result = $jobResult.results
+            Context "Verify spelling in $file" {
```
Comment (iSazonov) [inline test/common/markdown/markdown-spell.tests.ps1]: I think we need one It-block per file.

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+                $passes = @($result) -like '*free of spelling*'
+ 
+                $trueFailures = foreach ($Failure in $Failures) {
+                    @{ Spell = $Failure }
+                }
+
+                if($trueFailures) {
+                    it "<spell> should work" -TestCases $trueFailures {
+                        param($spell)
+                        throw "Tool reported spelling as wrong."
```
Comment (iSazonov) [inline test/common/markdown/markdown-spell.tests.ps1]: If we want to see spell errors we could output its here. 
Replace It-block with Should

Comment (RDIL) [reply test/common/markdown/markdown-spell.tests.ps1]: vexx32 seems to have it under control

**Comment by RDIL** [comment]
@vexx32 here is with 2 files:
https://cirrus-ci.com/task/5078643473645568

**Comment by vexx32** [comment]
@RDIL this is what I have so far
image

The code, cleaned and refactored:

```powershell
if (@($Output).Count -gt 1) {
    # Trim last line containing only success/failure message and split into one block per file
    $ErrorList = $Output[0..($Output.Length - 2)] -split '(\n[^|]+\n)'

    foreach ($Entry in ($ErrorList )) {
        $Lines = $Entry -split '\r?\n'
        $File = $Lines[0].Trim()
        $MistakeList = $Lines | Select-Object -Skip 1
        foreach ($Mistake in $MistakeList) {
            if ($Mistake -match '(?<Line>[0-9]+) \| (?<Mistake>[\w ]+)') {
                $Matches.Remove(0)
                $Matches.Add('Filename', $File)

                # Output the object from the regex match with filename
                [PSCustomObject]$Matches
            }
        }
    }
}
```

If you run this on the `mdspell` output sequence you should get usable PSCustomObjects listing file, line number, and the text that it detects containing a misspelling. These you can put into your pester test and output how you wish to show exactly which files and where the problems are. 😄

**Comment by RDIL** [comment]
@vexx32 nice.  I'll try to add it whenever I have time.

File: test/common/markdown/markdown-spell.tests.ps1 (hunk 19 above)
Comment (RDIL) [reply test/common/markdown/markdown-spell.tests.ps1]: It is now set up like that

**Comment by RDIL** [comment]
i give up these spelling errors cannot be fixed

**Comment by vexx32** [comment]
@RDIL if you work in the code to parse the output more granularly it should point you right to each error, instead of you having to scan through each file to figure out what's going wrong 😄

**Comment by RDIL** [comment]
@vexx32 I'm aware of which sentence the errors are in, but I don't know which word.

**Comment by vexx32** [comment]
List the remaining lines here, then, I guess?

**Comment by TravisEz13** [comment]
This looks like a good start.  It would be nice if we could use this version of the tool (https://github.com/lukeapage/node-markdown-spellcheck/pull/114) so that pester could parse the results and report the actual line and word that failed.

**Comment by TravisEz13** [comment]
Sorry, I didn't read @vexx32 	's comments. Before posting that, just consider it as one idea.

**Comment by RDIL** [comment]
I'm going to close this until I am in a better mental state and can work on this.

**Comment by PoshChan-Staging** [comment]
@web-flow, your last commit had 3 failures in `PowerShell-CI-static-analysis`
### Verify Markdown Spelling.Verify spelling in /home/vsts/work/1/s/tools/performance/README.md.>> 8 spelling errors found in 1 file!
  Error: Tool reported spelling as wrong.
```stacktrace
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-spell.tests.ps1: line 64
64:                         throw "Tool reported spelling as wrong."

```
### Verify Markdown Spelling.Verify spelling in /home/vsts/work/1/s/docs/cmdlet-example/command-line-simple-example.md.>> 5 spelling errors found in 1 file!
  Error: Tool reported spelling as wrong.
```stacktrace
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-spell.tests.ps1: line 64
64:                         throw "Tool reported spelling as wrong."

```
### Verify Markdown Spelling.Verify spelling in /home/vsts/work/1/s/CHANGELOG.md.>> 32 spelling errors found in 1 file!
  Error: Tool reported spelling as wrong.
```stacktrace
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-spell.tests.ps1: line 64
64:                         throw "Tool reported spelling as wrong."

```

**Comment by RDIL** [comment]
thanks poshchan I had absolutely no idea the check I wrote is failing

File: test/common/markdown/markdown.tests.ps1 [hunk 21]
```diff
@@ -17,8 +17,8 @@ Describe 'Common Tests - Validate Markdown Files' -Tag 'CI' {
         {
             $NpmInstalled = "Installed"
             Write-Verbose -Message "NPM is checking Gulp is installed. This may take a few moments." -Verbose
-            start-nativeExecution { sudo npm install --silent }
-            start-nativeExecution { sudo npm install 'gulp@4.0.0' --silent }
+            start-nativeExecution { yarn }
+            start-nativeExecution { yarn add gulp@4.0.0 }
```
Comment (iSazonov) [inline test/common/markdown/markdown.tests.ps1]: Maybe gulp@4.0.1?

Comment (RDIL) [reply test/common/markdown/markdown.tests.ps1]: waiting for merge of #9466

**Comment by RDIL** [comment]
This makes me so mad because this code works but this one file has a spelling mistake and I don't know what word it is :[

**Comment by ghvanderweg** [comment]
@RDIL which file and line number?

**Comment by RDIL** [comment]
It says it somewhere in the static analysis results

**Comment by ghvanderweg** [comment]
3 files with a total of 57 errors.. but nothing jumps out at me either. I'll take a closer look.
```
    Context Verify spelling in /home/vsts/work/1/s/docs/cmdlet-example/command-line-simple-example.md
WARNING: >> 5 spelling errors found in 1 file
      [-] >> 5 spelling errors found in 1 file! 81ms
        RuntimeException: Tool reported spelling as wrong.
        at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-spell.tests.ps1: line 65

    Context Verify spelling in /home/vsts/work/1/s/CHANGELOG.md
WARNING: >> 44 spelling errors found in 1 file
      [-] >> 44 spelling errors found in 1 file! 25ms
        RuntimeException: Tool reported spelling as wrong.
        at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-spell.tests.ps1: line 65

    Context Verify spelling in /home/vsts/work/1/s/tools/performance/README.md
WARNING: >> 8 spelling errors found in 1 file
      [-] >> 8 spelling errors found in 1 file! 2ms
        RuntimeException: Tool reported spelling as wrong.
        at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-spell.tests.ps1: line 65
```

**Comment by ghvanderweg** [comment]
It looks like the file overrides in .spelling don't work like they're supposed to. If I remove the file override lines and just put everything in the global dictionary, it works fine.

**Comment by RDIL** [comment]
Interesting, I’ll give it a look.

**Comment by RDIL** [comment]
Umm it didn’t work 😢

**Review by ghvanderweg — COMMENTED** [review]
You left the file exception for CHANGELOG.md in, so now it thinks all the exception that follow are just for this file
edit: how do I leave a comment at a certain line? This is my first time reviewing a PR

**Comment by RDIL** [comment]
Next to the line number there should be a clickable space which gives you the ability to type in a message

**Comment by RDIL** [comment]
Holy cow it worked

**Review by iSazonov — COMMENTED** [review]
Great work!

**Comment by RDIL** [comment]
@iSazonov are you sure removing the throw is the right thing? That would make the tests pass...

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+            Context "Verify spelling in $file" {
+                $failures = @($result) -like '*spelling errors found in*'
+                $passes = @($result) -like '*free of spelling*'
+ 
+                $trueFailures = foreach ($Failure in $Failures) {
+                    @{ Spell = $Failure }
+                }
+
+                if($trueFailures) {
+                    it "<spell>!" -TestCases $trueFailures {
```
Comment (iSazonov) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
                    It "<spell>!" -TestCases $trueFailures {
```

**Comment by iSazonov** [comment]
@RDIL My request is not remove throw but replace with Should - I'd expect more friendly output. I can wrong but you could make tests and compare the output.

**Comment by RDIL** [comment]
I'm sorry, but I don't really understand what you mean by that.

File: test/common/markdown/markdown-spell.tests.ps1 (hunk 20 above)
Comment (TravisEz13) [reply test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
                        throw "Tool reported spelling as wrong.`n$spell"
```

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+                $passes = @($result) -like '*free of spelling*'
+ 
+                $trueFailures = foreach ($Failure in $Failures) {
+                    @{ Spell = $Failure }
+                }
+
+                if($trueFailures) {
+                    It "<spell>!" -TestCases $trueFailures {
+                        param($spell)
+                        Write-Warning $spell
```
Comment (TravisEz13) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
```

Comment (TravisEz13) [reply test/common/markdown/markdown-spell.tests.ps1]: This isn't needed if the change below is taken.

File: test/common/markdown/markdown-spell.tests.ps1 (hunk 20 above)
Comment (iSazonov) [reply test/common/markdown/markdown-spell.tests.ps1]: If $trueFailures contains many values we get many lines "Tool reported spelling as wrong".

File: .spelling [hunk 24]
```diff
@@ -828,7 +836,6 @@ CodeFormatter
 StyleCop
 SytzeAndr
 yashrajbharti
- - CHANGELOG.md
```
Comment (TravisEz13) [inline .spelling]: Why did you get rid of the per document sections?

Comment (RDIL) [reply .spelling]: @ghvanderweg

Comment (ghvanderweg) [reply .spelling]: The per document sections were not being processed correctly, causing false negatives. I don't know if this is a regression in mdspell or an error in the testscript.

Comment (TravisEz13) [reply .spelling]: cc @adityapatwardhan

**Comment by iSazonov** [comment]
@RDIL I haven't experience to review this in depth. And MSFT team is busy in the days :-( So waiting...

File: test/common/markdown/markdown-link.tests.ps1 [hunk 25]
```diff
@@ -6,14 +6,7 @@ Describe "Verify Markdown Links" {
...
             Write-Verbose "installing markdown-link-check ..." -Verbose
-            start-nativeExecution {
-                sudo yarn global add markdown-link-check@3.7.2
-            }
-        }
-
-        if(!(Get-Module -Name 'ThreadJob' -ListAvailable -ErrorAction SilentlyContinue))
-        {
-            Install-Module -Name ThreadJob -Scope CurrentUser
+            start-nativeExecution { sudo yarn global add markdown-link-check@3.7.2 }
```
Comment (rjmholt) [inline test/common/markdown/markdown-link.tests.ps1]:
```suggestion
            Start-NativeExecution { sudo yarn global add markdown-link-check@3.7.2 }
```

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 26]
```diff
@@ -0,0 +1,68 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+Describe "Verify Markdown Spelling" {
+    BeforeAll {
+        if(!(Get-Command -Name 'mdspell' -ErrorAction SilentlyContinue)) {
```
Comment (rjmholt) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
        if (!(Get-Command -Name 'mdspell' -ErrorAction SilentlyContinue)) {
```

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+        }
+
+        Get-Job | Remove-Job -Force
+    }
+
+    AfterAll {
+        Get-Job | Remove-Job -Force
+    }
+
+    $groups = Get-ChildItem -Path "$PSScriptRoot\..\..\..\*.md" -Recurse |
```
Comment (rjmholt) [inline test/common/markdown/markdown-spell.tests.ps1]: This script should be in the `BeforeAll` block -- Pester doesn't guarantee the semantics of things the occur in the body of the `Describe`

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+        Get-Job | Remove-Job -Force
+    }
+
+    $groups = Get-ChildItem -Path "$PSScriptRoot\..\..\..\*.md" -Recurse | 
+        Where-Object DirectoryName -notlike '*node_modules*' | 
+        Group-Object -Property directory
+
+    $jobs = @{}
+    # Start all spelling verification in parallel
+    Foreach($group in $groups)
```
Comment (rjmholt) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
    foreach ($group in $groups)
```

Comment (rjmholt) [reply test/common/markdown/markdown-spell.tests.ps1]: This could now be `ForEach-Object -Parallel`

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+                    file = $file
+                    results = $results
+                })
+            }
+        } -ArgumentList @($group.Group)
+        $jobs.add($group.name,$job)
+    }
+
+    # Get the results and verify
+    foreach($key in $jobs.GetEnumerator().Name)
```
Comment (rjmholt) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
    foreach ($key in $jobs.GetEnumerator().Name)
```

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        $job = $jobs.$key
+        $results = Receive-Job -Job $job -Wait
+        Remove-job -Job $Job
+        foreach($jobResult in $results)
```
Comment (rjmholt) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
        foreach ($jobResult in $results)
```

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+            $result = $jobResult.results
+            Context "Verify spelling in $file" {
+                $failures = @($result) -like '*spelling errors found in*'
+                $passes = @($result) -like '*free of spelling*'
+ 
+                $trueFailures = foreach ($Failure in $Failures) {
+                    @{ Spell = $Failure }
+                }
+
+                if($trueFailures) {
```
Comment (rjmholt) [inline test/common/markdown/markdown-spell.tests.ps1]:
```suggestion
                if ($trueFailures) {
```

File: test/common/markdown/markdown-spell.tests.ps1 [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        $job = $jobs.$key
+        $results = Receive-Job -Job $job -Wait
+        Remove-job -Job $Job
+        foreach($jobResult in $results)
+        {
+            $file = $jobResult.file
+            $result = $jobResult.results
+            Context "Verify spelling in $file" {
+                $failures = @($result) -like '*spelling errors found in*'
```
Comment (rjmholt) [inline test/common/markdown/markdown-spell.tests.ps1]: There shouldn't be any need for the array notation here

Comment (RDIL) [reply test/common/markdown/markdown-spell.tests.ps1]: What do you mean?  I'm still a beginner to powershell so I'm looking to improve this any way I can

Comment (TravisEz13) [reply test/common/markdown/markdown-spell.tests.ps1]: `@()` means invoke this and return an array.  results should always be an array.  So there is no, need for this extra invocation.
```suggestion
                $failures = $result -like '*spelling errors found in*'
```
